src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.HandleDictionary.cs

```diff
@@ -0,0 +1,141 @@
+
+                if (enumerable == null)
+                {
+                    // Write a null object or enumerable.
+                    writer.WriteNull(jsonPropertyInfo.Name);
+                    return true;
+                }
+
+                state.Current.Enumerator = enumerable.GetEnumerator();
+
+                if (jsonPropertyInfo.Name == null)
+                {
+                    writer.WriteStartObject();
+                }
+                else
+                {
+                    writer.WriteStartObject(jsonPropertyInfo.Name);
+                }
+            }
+
+            if (state.Current.Enumerator.MoveNext())
+            {
+                // Check for polymorphism.
+                if (elementClassInfo.ClassType == ClassType.Unknown)
+                {
+                    //todo:test
+                    object currentValue = ((IDictionaryEnumerator)(state.Current.Enumerator)).Entry;
+                    GetRuntimeClassInfo(currentValue, ref elementClassInfo, options);
+                }
+
+                if (elementClassInfo.ClassType == ClassType.Value)
+                {
+                    elementClassInfo.GetPolicyProperty().WriteDictionary(options, ref state.Current, writer);
+                }
+                else if (state.Current.Enumerator.Current == null)
+                {
+                    writer.WriteNull(jsonPropertyInfo.Name);
+                }
+                else
+                {
+                    // An object or another enumerator requires a new stack frame.
+                    object nextValue = state.Current.Enumerator.Current;
+                    state.Push(elementClassInfo, nextValue);
+                }
+
+                return false;
+            }
+
+            // We are done enumerating.
+            writer.WriteEndObject();
+
+            state.Current.EndDictionary();
+
+            return true;
+        }
+
+        internal static void WriteDictionary<TProperty>(
+            JsonValueConverter<TProperty> converter,
+            JsonSerializerOptions options,
+            ref WriteStackFrame current,
+            Utf8JsonWriter writer)
+        {
+            if (converter == null)
+            {
+                return;
+            }
+
+            Debug.Assert(current.Enumerator != null);
+
+            string key;
+            TProperty value;
+            if (current.Enumerator is IEnumerator<KeyValuePair<string, TProperty>> enumerator)
+            {
+                // Avoid boxing for strongly-typed enumerators such as returned from IDictionary<string, TRuntimeProperty>
+                value = enumerator.Current.Value;
+                key = enumerator.Current.Key;
+            }
+            else
+            {
+                // Todo: support non-generic Dictionary here (IDictionaryEnumerator)
+                throw new NotSupportedException();
+            }
+
+            if (value == null)
+            {
+                writer.WriteNull(key);
```

Comment:
Should this go through converter as well just like the else path?

Reply:
The converters don't deal with null values so we do that here.

Reply:
Also the converters use (or will use) the already-escaped property name unlike the call to `WriteNull(key)` which will escape automatically. Dictionary key values can't be cached unlike property names.